### PR TITLE
Fix the mid-level pass pipeline

### PR DESCRIPTION
**Pull request by atrick**
This is a small series of commits that together fixes issues in the pass pipeline that are causing severe instability in benchmark performance and blocking other bug fixes.

We have heroic optimizations for array access, but whether those actually kick in was essentially random. Now, those optimizations always have a chance to run and the pass pipeline consistently reruns when necessary as it was designed to do.

The two main bug fixes are
- Don't inline array semantic calls before the mid-level pipeline
- Fix the mid-level pipeline restart mechanism

### CODE REVIEW

**Comment by atrick** [comment]
This PR is on hold because there are a few benchmark regressions I don't have time to investigate today. The vast majority are improvements, but I should still understand why the few regressions happened before merging.

**Comment by atrick** [comment]
This PR is still active. I plan to work on merging it this week after finishing some work on EscapeAnalysis. Unfortunately, I can't avoid all regressions because of some fundamental problems in ARC optimization.

File: test/SILOptimizer/array_contentof_opt.swift
Comment (eeckstein) [inline test/SILOptimizer/array_contentof_opt.swift]: The purpose of this test is to check if the array.append(contentof) optimization works.
With your change it does not work anymore. Adding this "workaround" option to let the test pass does not make any sense.

**Review by eeckstein — CHANGES_REQUESTED** [review]

File: test/SILOptimizer/array_contentof_opt.swift
Comment (atrick) [reply test/SILOptimizer/array_contentof_opt.swift]: @eeckstein 
As the commit log said:
fix a test case that assumes Array.append is never inlined.

I just added another comment to make it more clear:
// The checks lines are written assuming that the array(Element) calls
// are never inlined. Consequently, the RUN line must prevent inlining
// Array(Element).

What makes you think this PR broke an optimization?

Comment (eeckstein) [reply test/SILOptimizer/array_contentof_opt.swift]: As written in the comment in the test file: This is an end-to-end test of the array(contentsOf) -> array(Element) optimization

This optimization replaces an Array.append(contentof:) with a sequence of Array.append(element).
The point of this test is to check if the optimization works. And working means: out of the box without any special command line options.

Comment (atrick) [reply test/SILOptimizer/array_contentof_opt.swift]: @eeckstein I don't understand why you think the optimization is broken. If array(Element) is inlined after the optimization, then the test's CHECK lines will fail. So it wouldn't be an effective end-to-end test. A proper end-to-end test is going to need to control inlining if it's checking for the presence of calls.

Comment (eeckstein) [reply test/SILOptimizer/array_contentof_opt.swift]: Sorry, I misunderstood the change here. Looks good!

**Review by eeckstein — APPROVED** [review]
lgtm

File: test/SILOptimizer/array_contentof_opt.swift
Comment (atrick) [reply test/SILOptimizer/array_contentof_opt.swift]: I cleared up the test case comments
https://github.com/apple/swift/pull/22445/commits/1e73e548caabf8bcf8d598105adadaea67f4d594

Thanks

**Comment by atrick** [comment]
@swift-ci test

**Comment by atrick** [comment]
@swift-ci benchmark

**Comment by atrick** [comment]
@swift-ci test source compatibility

**Comment by atrick** [comment]
@swift-ci test compiler performance

**Comment by atrick** [comment]
@swift-ci test compiler performance

**Comment by atrick** [comment]
@swift-ci test compiler performance

**Comment by futurejones** [comment]
Can one of the admins verify this patch?

**Comment by atrick** [comment]
Superseded by #31163 and other followup PRs.

**Pull request closed without merging**
